lib/segment/src/index/field_index/numeric_index/immutable_numeric_index.rs

```diff
@@ -0,0 +1,537 @@
+use std::collections::BTreeMap;
+use std::ops::Bound::{Excluded, Unbounded};
+use std::ops::{Bound, Range};
+use std::sync::Arc;
+
+use common::types::PointOffsetType;
+use parking_lot::RwLock;
+use rocksdb::DB;
+
+use super::mutable_numeric_index::MutableNumericIndex;
+use super::{Encodable, NumericIndex, HISTOGRAM_MAX_BUCKET_SIZE, HISTOGRAM_PRECISION};
+use crate::common::operation_error::OperationResult;
+use crate::common::rocksdb_wrapper::DatabaseColumnWrapper;
+use crate::index::field_index::histogram::{Histogram, Numericable, Point};
+
+pub struct ImmutableNumericIndex<T: Encodable + Numericable> {
+    map: NumericKeySortedVec<T>,
+    db_wrapper: DatabaseColumnWrapper,
+    pub(super) histogram: Histogram<T>,
+    pub(super) points_count: usize,
+    pub(super) max_values_per_point: usize,
+    point_to_values: Vec<Range<u32>>,
+    point_to_values_container: Vec<T>,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub(super) struct NumericIndexKey<T> {
+    key: T,
+    idx: PointOffsetType,
+    deleted: bool,
+}
+
+struct NumericKeySortedVec<T: Encodable + Numericable> {
+    data: Vec<NumericIndexKey<T>>,
+    deleted_count: usize,
+}
+
+pub(super) struct NumericKeySortedVecIterator<'a, T: Encodable + Numericable> {
+    set: &'a NumericKeySortedVec<T>,
+    start_index: usize,
+    end_index: usize,
+}
+
+impl<T: PartialEq + PartialOrd + Encodable + Numericable> NumericIndexKey<T> {
+    pub(super) fn new(key: T, idx: PointOffsetType) -> Self {
+        Self {
+            key,
+            idx,
+            deleted: false,
+        }
+    }
+
+    pub(super) fn encode(&self) -> Vec<u8> {
+        T::encode_key(&self.key, self.idx)
+    }
+
+    pub(super) fn decode(bytes: &[u8]) -> Self {
+        let (idx, key) = T::decode_key(bytes);
+        Self {
+            key,
+            idx,
+            deleted: false,
+        }
+    }
+}
+
+impl<T> From<NumericIndexKey<T>> for Point<T> {
+    fn from(key: NumericIndexKey<T>) -> Self {
+        Point {
+            val: key.key,
+            idx: key.idx as usize,
+        }
+    }
+}
+
+impl<T> From<Point<T>> for NumericIndexKey<T> {
+    fn from(key: Point<T>) -> Self {
+        NumericIndexKey {
+            key: key.val,
+            idx: key.idx as PointOffsetType,
+            deleted: false,
+        }
+    }
+}
+
```

Comment:
@IvanPleshkov  I am running Rust 1.73 locally and Clippy complains about this comparator.

```
error: incorrect implementation of `partial_cmp` on an `Ord` type
  --> lib/segment/src/index/field_index/numeric_index/immutable_numeric_index.rs:86:1
   |
86 | /  impl<T: PartialEq + PartialOrd + Encodable> PartialOrd for NumericIndexKey<T> {
87 | |      fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
   | | _______________________________________________________________________-
88 | ||         Some(match self.key.cmp_encoded(&other.key) {
89 | ||             std::cmp::Ordering::Equal => self.idx.cmp(&other.idx),
90 | ||             ord => ord,
91 | ||         })
92 | ||     }
   | ||_____- help: change this to: `{ Some(self.cmp(other)) }`
93 | |  }
   | |__^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#incorrect_partial_ord_impl_on_ord_type
   = note: `#[deny(clippy::incorrect_partial_ord_impl_on_ord_type)]` on by default
```

Reply:
FYI applying the proposed fix is not great

```
thread 'index::field_index::numeric_index::immutable_numeric_index::tests::test_numeric_index_key_set' has overflowed its stack
fatal runtime error: stack overflow
```

Reply:
This fix is implemented in <https://github.com/qdrant/qdrant/pull/2767>.